pkg/pillar/cmd/zedagent/handlemetrics.go

```diff
@@ -174,22 +174,57 @@ func metricsTimerTask(ctx *zedagentContext, handleChannel chan interface{}) {
+// This method handles 2 scenarios:
+// 1. Current metrics publish interval > new metrics publish interval in GlobalConfig
+//     For instance, if currentMetricInterval = 300s and latestMetricTickerInterval = 200s,
+//     the controller will be expecting a metrics every 300s. So increasing the frequency
+//     immediately to 200s will not result in controller marking the edge-node as suspect
+//     in between.
+// 2. Current metrics publish interval < new metrics publish interval in GlobalConfig
+//     For instance, if currentMetricInterval = 200s and latestMetricTickerInterval = 300s,
+//     the controller will be expecting a metrics every 200s. So decreasing the frequency to
+//     300s can result in controller marking the edge-node as suspect in between. To avoid this,
+//     we have to make sure that the controller is notified  to expect a metrics every 300s before
+//     updating our publish frequency. forceUpdate should be true only after we've successfully
+//     notified new frequency to the controller.
+func maybeUpdateMetricsTimer(ctx *getconfigContext, forceUpdate bool) {
+	latestMetricsInterval := ctx.zedagentCtx.globalConfig.GlobalValueInt(types.MetricInterval)
+	log.Functionf("maybeUpdateMetricsTimer: currentMetricInterval %v, latestMetricsInterval %v, forceUpdate: %v",
+		ctx.currentMetricInterval, latestMetricsInterval, forceUpdate)
+	if ctx.currentMetricInterval > latestMetricsInterval {
+		log.Tracef("maybeUpdateMetricsTimer: updating metrics publish interval %v seconds",
+			latestMetricsInterval)
+		updateMetricsTimer(ctx, latestMetricsInterval)
+	} else if ctx.currentMetricInterval < latestMetricsInterval {
+		if forceUpdate {
+			log.Tracef("maybeUpdateMetricsTimer: updating metrics publish interval %v seconds",
+				latestMetricsInterval)
+			updateMetricsTimer(ctx, latestMetricsInterval)
+		} else {
+			// Force an immediate timeout to publish metrics.
+			flextimer.TickNow(ctx.metricsTickerHandle)
+		}
+	}
+}
+
 // Called when globalConfig changes
 // Assumes the caller has verifier that the interval has changed
-func updateMetricsTimer(metricInterval uint32, tickerHandle interface{}) {
+func updateMetricsTimer(ctx *getconfigContext, metricInterval uint32) {
 
-	if tickerHandle == nil {
+	if ctx.metricsTickerHandle == nil {
 		log.Warnf("updateMetricsTimer: no metricsTickerHandle yet")
 		return
 	}
 	interval := time.Duration(metricInterval) * time.Second
 	log.Functionf("updateMetricsTimer() change to %v", interval)
 	max := float64(interval)
 	min := max * 0.3
-	flextimer.UpdateRangeTicker(tickerHandle,
+	flextimer.UpdateRangeTicker(ctx.metricsTickerHandle,
 		time.Duration(min), time.Duration(max))
-	// Force an immediate timout since timer could have decreased
-	flextimer.TickNow(tickerHandle)
+	ctx.currentMetricInterval = metricInterval
+	// Force an immediate timeout since timer could have decreased
+	flextimer.TickNow(ctx.metricsTickerHandle)
```

Comment:
This comment and code seems to be misplaced now; you're doing it in the new Maybe function AFAICT.

Reply:
I left that part as it is in `updateMetricsTimer` method because in `maybeUpdateMetricsTimer` method I'm firing the ticker only under a certain condition. But in `updateMetricsTimer` method we might have to always send out a metric to reset the controller's timer. For example, if the dormant check interval on the controller is for 10 min and in between that interval at 8th min I update the timer value to 5 min. Upon receiving that update EVE must immediately send out a metric to reset the controller's timer so that it doesn't mark the device as a suspect after the 10th min.